File: backend/app/models/workspace.py
import uuid
from enum import Enum
from datetime import datetime, timezone
from sqlalchemy import String, Enum as SQLEnum, ForeignKey, UniqueConstraint, DateTime
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column
from app.models.base import BaseModel

class WorkspaceRole(str, Enum):
    owner = "owner"
    admin = "admin"
    member = "member"
    viewer = "viewer"

class WorkspaceMemberStatus(str, Enum):
    active = "active"
    suspended = "suspended"

class InvitationStatus(str, Enum):
    pending = "pending"
    accepted = "accepted"
    expired = "expired"

class Workspace(BaseModel):
    __tablename__ = "workspaces"

    name: Mapped[str] = mapped_column(String, nullable=False)
    owner_uid: Mapped[str] = mapped_column(String, index=True, nullable=False)  # Firebase UID của Owner

class WorkspaceMember(BaseModel):
    __tablename__ = "workspace_members"

    workspace_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("workspaces.id", ondelete="CASCADE"),
        index=True,
        nullable=False
    )
    user_uid: Mapped[str] = mapped_column(
        String,
        ForeignKey("users.firebase_uid", ondelete="CASCADE"),
        index=True,
        nullable=False
    )
    role: Mapped[WorkspaceRole] = mapped_column(
        SQLEnum(WorkspaceRole),
        default=WorkspaceRole.member,
        nullable=False
    )
    status: Mapped[WorkspaceMemberStatus] = mapped_column(
        SQLEnum(WorkspaceMemberStatus),
        default=WorkspaceMemberStatus.active,
        nullable=False
    )

    __table_args__ = (
        UniqueConstraint("workspace_id", "user_uid", name="uq_workspace_member"),
    )

class WorkspaceInvitation(BaseModel):
    __tablename__ = "workspace_invitations"

    workspace_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("workspaces.id", ondelete="CASCADE"),
        index=True,
        nullable=False
    )
    email: Mapped[str] = mapped_column(String, index=True, nullable=False)
    role: Mapped[WorkspaceRole] = mapped_column(
        SQLEnum(WorkspaceRole),
        default=WorkspaceRole.member,
        nullable=False
    )
    token: Mapped[str] = mapped_column(String, unique=True, index=True, nullable=False)
    invited_by: Mapped[str] = mapped_column(
        String,
        ForeignKey("users.firebase_uid", ondelete="CASCADE"),
        nullable=False
    )
    status: Mapped[InvitationStatus] = mapped_column(
        SQLEnum(InvitationStatus),
        default=InvitationStatus.pending,
        nullable=False
    )
    expires_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)

File: frontend/src/features/settings/types/workspace.ts
import { z } from "zod";

export const workspaceSchema = z.object({
  id: z.string().min(1),
  name: z.string(),
  owner_uid: z.string(),
  created_at: z.string().nullable().optional().default(null),
});

export type Workspace = z.infer<typeof workspaceSchema>;

export const workspaceRoleSchema = z.enum([
  "owner",
  "admin",
  "member",
  "viewer",
]);
export const workspaceMemberStatusSchema = z.enum(["active", "suspended"]);
export const workspaceInvitationStatusSchema = z.enum([
  "pending",
  "accepted",
  "expired",
]);

export const workspaceMemberSchema = z.object({
  workspace_id: z.string().min(1),
  user_uid: z.string().min(1),
  email: z.string().email().nullable(),
  full_name: z.string().nullable(),
  avatar_url: z.string().nullable(),
  role: workspaceRoleSchema,
  status: workspaceMemberStatusSchema,
});

export const workspaceInvitationSchema = z.object({
  id: z.string().min(1),
  workspace_id: z.string().min(1),
  email: z.string().email(),
  role: workspaceRoleSchema,
  token: z.string().min(1),
  invited_by: z.string().min(1),
  status: workspaceInvitationStatusSchema,
  expires_at: z.string().min(1),
  created_at: z.string().min(1),
});

export const workspaceOauthUrlSchema = z.object({
  oauth_url: z.string().url(),
});

export type WorkspaceRole = z.infer<typeof workspaceRoleSchema>;
export type WorkspaceMemberStatus = z.infer<typeof workspaceMemberStatusSchema>;
export type WorkspaceMember = z.infer<typeof workspaceMemberSchema>;
export type WorkspaceInvitation = z.infer<typeof workspaceInvitationSchema>;
export type WorkspaceOauthUrl = z.infer<typeof workspaceOauthUrlSchema>;

export interface WorkspaceNamePayload {
  name: string;
}
export interface WorkspaceInvitationPayload {
  email: string;
  role: WorkspaceRole;
}
export interface WorkspaceMemberUpdatePayload {
  role: WorkspaceRole;
  status: WorkspaceMemberStatus;
}
export interface WorkspaceOauthUrlPayload {
  platform: string;
  redirect_uri?: string;
}
export interface WorkspaceOauthCallbackPayload {
  code: string;
  state?: string;
  platform?: string;
}
